hutch_python/obj_config.py

```diff
@@ -0,0 +1,189 @@
+) -> None:
+    """
+    Helper function for updating object namespaces
+    obj_ns is a HelpfulNamespace, which may include other
+    HelpfulNamespace's.  As a result this must be semi-recursive.
+
+    Parameters
+    ----------
+    obj_ns : HelpfulNamespace
+        Namespace of all objects
+
+    entry : str
+        Entry in obj_conf yaml. Name of device or class of devices to edit.
+
+    attrs : list[str]
+        List of attributes to add to whitelist
+
+    fn : Callable[[BaseInterface, dict], None]
+        A function that applies a change to a singular device
+    """
+    dev = getattr(obj_ns, entry, None)
+    if dev:  # look for explicit device
+        fn(dev, attrs)
+
+        return
+
+    # look for devices with class
+    for dev in obj_ns:
+        if isinstance(dev, utils.HelpfulNamespace):
+            update_objs(dev, entry, attrs, fn)
+        # a poor-man's type check, in an effort to avoid importing
+        # every device type again
+        if entry == type(dev).__name__:
+            fn(dev, attrs)
+
+
+def update_whitelist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Update the tab whitelist for ``dev`` with each field in ``attrs``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to add to ``dev`` tablist
+    """
+    for att in attrs:
+        dev._tab.add(att)
+
+
+def update_blacklist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Remove each field in ``attrs`` from tab whitelist for ``dev``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to remove from ``dev`` tablist
+    """
+    for att in attrs:
+        try:
+            dev._tab.remove(att)
+        except KeyError:
+            logger.debug(f'key: {att} not in tab completion list')
+
+
+def replace_tablist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Replace tab completion list for ``dev`` with ``attrs``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to replace ``dev`` tablist with
+    """
+    for old_att in dir(dev):
+        dev._tab.remove(old_att)
+
+    for att in attrs:
+        dev._tab.add(att)
+
+
+def update_kind(dev: BaseInterface, attrs: dict) -> None:
+    """
+    Update the kind of components in ``dev``.  If a key matches the
+    device name, the top-level device kind will be modified.
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
```

Comment:
A bit pedantic of me, but I think the description here is too narrow. This  kind updater should work on any ophyd device, right?

Reply:
If you mean just for `update_kind`, you're absolutely right.